src/main/java/life/qbic/model/download/QbicDataFinder.java

```diff
@@ -62,7 +66,7 @@ public QbicDataFinder(
     Map<String, List<DataSet>> sampleDatasets = new HashMap<>();
     sampleDatasets.put(sample.getCode(), sample.getDataSets());
     for (Sample child : children) {
-      Map<String, List<DataSet>> childDatasetMapping = fetchDescendantDatasets(child);
+      Map<String, List<DataSet>> childDatasetMapping = fetchDescendantDatasets(child,visitedSamples);
```

Comment:
What do you think about passing down the map itself? Instead of the `put` method the `putIfAbsent` method could be used then.

Reply:
I am not sure if that would work. For each recursion you start you visit different samples. You need to remember where you started (findAllDatasetsRecursive) and what children you have visted (fetchDescendantDatasets).

We could also completely rewrite the code and just start in the "highest" point in our openbis structure and iterate through all children. This would prevent from iterating multiple times over the same samples. But then we would need to assume that all samples are attached in the correct manner.

Reply:
do we have a root? if so, I think this would be best.